File: demo/csep590b_c04/p3_count_groups/count_groups_fixed.cpp
/*
DEMO: Problem 3 (count_groups) — FIXED reference.

Corrected version of the buggy algorithm:

  * Always reset `run_len = 0` when a 0-bit closes a run, regardless
    of whether the closing run was counted.
  * After the loop, flush a trailing run-of-length-2+ that ran into
    bit 31 without being closed by a 0-bit.

Loop bound is the compile-time constant 32, so SAW fully unrolls.

Expected verdict: SAT (VERIFIED) against `count_groups_spec`.
*/

#include <cstdint>

uint32_t count_groups(uint32_t x) {
    uint32_t cnt = 0;
    uint32_t in_run = 0;
    uint32_t run_len = 0;
    for (int i = 0; i < 32; i++) {
        uint32_t bit = (x >> i) & 1u;
        if (bit == 1u) {
            in_run = 1u;
            run_len = run_len + 1u;
        } else {
            if (in_run == 1u) {
                if (run_len >= 2u) {
                    cnt = cnt + 1u;
                }
                in_run = 0u;
                run_len = 0u;     // fix: always reset
            }
        }
    }
    // fix: flush any trailing run that extended through bit 31
    if (in_run == 1u && run_len >= 2u) {
        cnt = cnt + 1u;
    }
    return cnt;
}

File: demo/csep590b_c04/p3_count_groups/count_groups_fixed.rs
// DEMO: Problem 3 (count_groups) — FIXED reference (Rust mirror of
// count_groups_fixed.cpp).
//
// Always resets run_len, and flushes a trailing run that extends
// through bit 31.
//
// Expected verdict: VERIFIED against `count_groups_spec`.

fn count_groups(x: u32) -> u32 {
    let mut cnt: u32 = 0;
    let mut in_run: u32 = 0;
    let mut run_len: u32 = 0;
    let mut i: u32 = 0;
    while i < 32 {
        let bit = (x >> i) & 1;
        if bit == 1 {
            in_run = 1;
            run_len = run_len.wrapping_add(1);
        } else if in_run == 1 {
            if run_len >= 2 {
                cnt = cnt.wrapping_add(1);
            }
            in_run = 0;
            run_len = 0;
        }
        i = i.wrapping_add(1);
    }
    if in_run == 1 && run_len >= 2 {
        cnt = cnt.wrapping_add(1);
    }
    cnt
}
